[offload] - Remove standalone build in favor of 'runtimes'

estewart08 opened the pull request:
Summary:
Follow up on removal of OPENMP_STANDALONE_BUILD in openmp (#149878). This
build method is redundant and can be accomplished via runtimes.

Removes support for:
`cmake -S <llvm-project>/offload ...`

 Switches over to:
 `make -S <llvm-project>/runtimes -DLLVM_ENABLE_RUNTIMES=openmp;offload ...`

Libomptarget has a dependency on libomp.so and requires the omp cmake target to exist at build time, which is why both runtimes are listed.

Updates cmake compiler logic in offload/CMakeLists.txt to mirror openmp changes:
 [openmp] Allow testing OpenMP without a full clang build tree (#182470)

User will still need to have a separate invocation to build openmp DeviceRTL via:
`-DLLVM_ENABLE_RUNTIMES=openmp`
`-DLLVM_DEFAULT_TARGET_TRIPLE=<amdgcn-amd-amdhsa|nvptx64-nvidia-cuda>`

jhuber6 reviewed (commented):
Thanks for the work here! Just one concern, and also we should probably try to be consistent with `if ()` vs. `if()`. I think we prefer the latter.

jprotze commented:
For a standalone build, it was sufficient to download the openmp src tarball and build the code: https://github.com/llvm/llvm-project/releases/tag/llvmorg-21.1.7

The documentation should explain, which source tarballs are necessary and sufficient for the new workflow and how we need to extract them

mgorny commented:
How is it supposed to work? FWICS `llvm_gtest` target requires `find_package(LLVM)` to succeed but the runtimes CMakeLists has only:

```
find_package(LLVM PATHS "${LLVM_BINARY_DIR}" NO_DEFAULT_PATH NO_CMAKE_FIND_ROOT_PATH)
```

which cannot succeed because `LLVM_BINARY_DIR` is empty and `NO_DEFAULT_PATH` suppresses any way to specify the correct path.

estewart08 commented on offload/CMakeLists.txt:
```diff
@@ -55,77 +40,46 @@ list(INSERT CMAKE_MODULE_PATH 0
 set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${LIBOMPTARGET_LIBRARY_DIR})
 
-if(NOT LLVM_LIBRARY_OUTPUT_INTDIR)
+if(NOT LLVM_LIBRARY_OUTPUT_INTDIR OR NOT TARGET Clang)
```
Yes, the Clang target is present in the normal runtimes builds. It actually is in the deps list in `runtimes/CMakeLists.txt`. This prevents the standalone build from thinking the build tree is where llvm is currently installed.

shiltian commented:
I don’t think we should drop standalone builds. Many users of `libomp` don't care about offloading at all, so losing the ability to build it standalone would definitely be a step backward.

jhuber6 reviewed (approved):
I'm fine with removing both, it's much easier to constrain the different configurations we need to test. If anyone out there is still relying on a dedicated checkout of OpenMP they should either move to the new format (You need runtimes, llvm, and openmp instead of just openmp directories) or write some convincing arguments for why we should keep both of them.

Meinersbur commented:
> The only significant difference is that the first only requires the `openmp` directory while the other requires an LLVM checkout. I think it's reasonable to require an LLVM checkout for these always.

We have shared code [third-party/](https://github.com/llvm/llvm-project/tree/main/third-party) and [cmake/](https://github.com/llvm/llvm-project/tree/main/cmake). The former is used by openmp since #147381. The latter will be after #171515. While it is possible to make OpenMP work without support for those features, it adds complexity to the build.


> I recommend removing standalone build for `libomptarget`, and still keep `libomp` as it is. `libomp` doesn't depend on LLVM anyway, except those very limited feature(s) that are rarely used.

The primary motivation is reducing build complexity. We have 4 modes how to build OpenMP: 
 * `cmake <llvm-project>/llvm -DLLVM_ENABLE_PROJECTS=openmp` (#152189)
 * `cmake <llvm-project>/llvm -DLLVM_ENABLE_RUNTIMES=openmp` (bootstrapping build)
 * `cmake <llvm-project>/openmp` (standalone)
 * `cmake <llvm-project>/runtimes -DLLVM_ENABLE_RUNTIMES=openmp` (runtimes defalt build)

We should not have to maintain build modes which are redundant, not tested with out CI infrastructures, different in subtle ways, constantly break when touch a `CMakeLists.txt` and thus must be tested manually. Few contributors do which means openmp is broken in most modes (such as observed in this very PR). See https://discourse.llvm.org/t/rfc-llvm-build-system-future-direction/53430 for a previous effort on reducing the combinatorial explosion of build configurations.

mgorny commented:
…and that would be because all the logic for finding installed `openmp` library/headers is removed.

jhuber6 commented:
> …and that would be because all the logic for finding installed `openmp` library/headers is removed.

That'd do it, I suppose we still want to use `DetectTestCompiler` in this case? It could be either in a default runtimes build.

estewart08 committed:
[offload] - Remove standalone build in favor of 'runtimes'

Follow up on removal of OPENMP_STANDALONE_BUILD in openmp (#149878). This
build method is redundant and can be accomplished via runtimes.

Removes support for:
  cmake -S <llvm-project>/offload ...

Switches over to:
 make -S <llvm-project>/runtimes -DLLVM_ENABLE_RUNTIMES=openmp;offload ...

Libomptarget has a dependency on libomp.so and requires the omp cmake target
to exist at build time, which is why both runtimes are listed.

Updates cmake compiler logic in offload/CMakeLists.txt to mirror
openmp changes:
[openmp] Allow testing OpenMP without a full clang build tree (#182470)

User will still need to have a separate invocation to build openmp DeviceRTL via:
-DLLVM_ENABLE_RUNTIMES=openmp
-DLLVM_DEFAULT_TARGET_TRIPLE=<amdgcn-amd-amdhsa|nvptx64-nvidia-cuda>

Meinersbur commented on openmp/docs/SupportAndFAQ.rst:
```diff
@@ -34,6 +34,22 @@ OpenMP in Flang Technical Call
+     -DCMAKE_INSTALL_PREFIX=<PATH>        \ # Where the libraries will live
+  $> ninja install
+
+A separate cmake invocation to build the openmp deviceRTL via ``-DLLVM_ENABLE_RUNTIMES=openmp`` and ``-DLLVM_DEFAULT_TARGET_TRIPLE=<amdgcn-amd-amdhsa|nvptx64-nvidia-cuda>`` is needed.
```
Build instructions have been consolidated into https://github.com/llvm/llvm-project/blob/main/openmp/docs/Building.md. This should already been convered by https://openmp.llvm.org/Building.html#building-with-offload-support. If not, can you update it there?

Note:  a comment after `\` negates its function as a line continuation

Meinersbur reviewed (approved):
LGTM
